src/AutoMapper/QueryableExtensions/ProjectionBuilder.cs

```diff
@@ -76,6 +76,10 @@ Expression CreateProjection(in ProjectionRequest request, LetPropertyMaps letPro
         foreach(var derivedMap in polymorphicMaps)
         {
             var sourceType = derivedMap.SourceType;
+            if (!request.SourceType.IsAssignableFrom(sourceType))
+            {
+                continue;
+            }
             var derivedRequest = request.InnerRequest(sourceType, derivedMap.DestinationType);
             var derivedProjection = CreateProjectionCore(derivedRequest, letPropertyMaps, derivedMap, Convert(source, sourceType));
```

Comment:
I think the only change needed is to replace `Convert` with `TypeAs`.

Reply:
you are correct, there is a difference in the generated Expressions and SQL though:

my initial solution does not add polymorphic maps that are not applicable to the expression tree:

```
DbSet<Car>()
    .Select(dtoCar => new VehicleModel{ Name = dtoCar.Name }
    )
    
SELECT [v].[Name]
FROM [Vehicles] AS [v]
WHERE [v].[Discriminator] = N'Car'
```

the approach of adding them with the type check results in:

```
DbSet<Car>()
    .Select(dtoCar => (dtoCar is Bicycle) ? new BicycleModel{ 
        EBike = (dtoCar as Bicycle).EBike, 
        Name = (dtoCar as Bicycle).Name 
    }
     : (dtoCar is Motorcycle) ? new MotorcycleModel{ 
        HasSidecar = (dtoCar as Motorcycle).HasSidecar, 
        Name = (dtoCar as Motorcycle).Name 
    }
     : new VehicleModel{ Name = dtoCar.Name }
    )

SELECT [v].[Id], [v].[Discriminator], [v].[Name], [v].[AmountDoors]
FROM [Vehicles] AS [v]
WHERE [v].[Discriminator] = N'Car'
```

if this is not an issue i can commit the version you proposed with less changes to the codebase

Reply:
ok, i cleaned up the commits to reflect the requested changes